### ISSUE TITLE
Proposed change to record segmentation

### ISSUE THREAD

**Issue by justinlittman** [iipc/warc-specifications#28]
The record segmentation states:
```
The WARC-Payload-Digest recorded in the first segment of a segmented record is the digest of the payload of the logical record.
```

This requires that a writer have the entire logical record before being able to write the first segment; in the case of really large payloads (which record segmentation is intended to support), this may be problematic.

My recommendation is to move this to last segment.

As some additional background:  As part of the [Social Feed Manager project](https://github.com/gwu-libraries/sfm-ui), we're working on recording the [Twitter streaming API](https://dev.twitter.com/streaming/overview) to WARCs.  As part of the streaming API, the HTTP response is kept open for an extended period of time (as in hours or days).

**Comment by saraaubry** [comment]
The following changes have been integrated in the revised ISO draft during the ISO working group meeting on November 16-17, 2015 :

in section 5.9, change the sentence "The WARC-Payload-Digest recorded in the first segment of a segmented record is the digest of the payload of the logical record." to "The WARC-Payload-Digest recorded in the segments of a segmented record is the digest of the payload of the logical record."

+ add a WARC-Payload-Digest in the B.8 continuation record example.

If you don't have the payload on the first segment, don't use this filed.

**Comment by justinlittman** [comment]
What does this mean (and can it be stated more clearly)?  "The WARC-Payload-Digest recorded in the segments of a segmented record is the digest of the payload of the logical record."

**Comment by kris-sigur** [comment]
@justinlittman It means that the `WARC-Payload-Digest` refers to the entire (reassembled) record/file (that may be spread over any number of WARC segment records).

**Comment by justinlittman** [comment]
That for the clarification.  I guess I'm confused by "recorded in the segments of a segmented record" (but not "The WARC-Payload-Digest is the digest of the payload of the logical record.")

Which segment should it appear in?  First?  Last?  Any?  Every?  Some?  Multiple?  What if it conflicts?

**Comment by kris-sigur** [comment]
@justinlittman The first one. 

> The WARC-Payload-Digest recorded in the first segment of a segmented record is the digest of the payload of the logical record.

This is all cover in [Chapter 7: Record Segmentation](https://iipc.github.io/warc-specifications/specifications/warc-format/warc-1.0/#record-segmentation)

**Comment by justinlittman** [comment]
Note that that the motivation for this proposed change is because recording it in the first segment is problematic.  See the first comment in this ticket.

**Comment by kris-sigur** [comment]
My apologies, I'd incorrectly remembered the exact history of this issue.

To be honest, I'm not sure a change is necessary here.

The `WARC-Payload-Digest` is an optional field. You could simple omit it in the scenario you describe in the initial post. 

To be able to detect bitrot you can then use the [WARC-Block-Digest](https://iipc.github.io/warc-specifications/specifications/warc-format/warc-1.0/#warc-block-digest).

You can then either add a metadata record with the final digest, add it to the final segment using a non-standard header or let downstream tools calculate it as needed. 

The change proposal @saraaubry entered seems fraught with danger, notably of collisions as @justinlittman has pointed out. It also risks confusing it and WARC-Block-Digest. As it stands it should not be adopted.

At minimum it should be stated that if there are more than one segment with a `WARC-Payload-Digest` field they must all be equal. If they are not equal the all the relevant records should be regarded as invalid. Alternatively that only one of the segments may have this field. 

It might also make sense to only allow this on the first and last segment. 

However it is very hard to validate any of these rules as doing so requires scanning multiple WARCs.

**Comment by ato** [comment]
@JustAnotherArchivist, yes, I agree with your reading. The change in 1.1 removed the guidance on which segment should have the field. Note that is what Justin and Kris were both objecting to in the comments above but I guess it passed anyway.
